--- win32/InputCustom.cpp
+++ win32/InputCustom.cpp
@@ -115,166 +115,48 @@
 
 void FunkyJoyStickTimer()
 {
-JOYINFOEX jie;
+S9xWinScanJoypads();
 
     for (short C = 0; C != 16; C ++)
     {
-
-        jie.dwSize = sizeof( jie);
-        jie.dwFlags = JOY_RETURNALL;
-
-        if (joyGetPosEx (JOYSTICKID1 + C, &jie) != JOYERR_NOERROR)
-            continue;
-
-        CheckAxis (C, 0, jie.dwXpos,
-                   JoystickF[C].Caps.wXmin, JoystickF[C].Caps.wXmax,
-                   JoystickF[C].Left, JoystickF[C].Right);
-        CheckAxis (C, 2, jie.dwYpos,
-                   JoystickF[C].Caps.wYmin, JoystickF[C].Caps.wYmax,
-                   JoystickF[C].Up, JoystickF[C].Down);
-        if(JoystickF[C].Caps.wCaps & JOYCAPS_HASZ)
-{
-CheckAxis (C, 41, jie.dwZpos,
-                   JoystickF[C].Caps.wZmin, JoystickF[C].Caps.wZmax,
-                   JoystickF[C].ZUp, JoystickF[C].ZDown);
-}
-if(JoystickF[C].Caps.wCaps & JOYCAPS_HASR)
-{
-        CheckAxis (C, 43, jie.dwRpos,
-                   JoystickF[C].Caps.wRmin, JoystickF[C].Caps.wRmax,
-                   JoystickF[C].RUp, JoystickF[C].RDown);
-}
-if(JoystickF[C].Caps.wCaps & JOYCAPS_HASU)
-{
-        CheckAxis (C, 45, jie.dwUpos,
-                   JoystickF[C].Caps.wUmin, JoystickF[C].Caps.wUmax,
-                   JoystickF[C].UUp, JoystickF[C].UDown);
-}
-        if(JoystickF[C].Caps.wCaps & JOYCAPS_HASV)
-{
-CheckAxis (C, 47, jie.dwVpos,
-                   JoystickF[C].Caps.wVmin, JoystickF[C].Caps.wVmax,
-                   JoystickF[C].VUp, JoystickF[C].VDown);
-}
-
-        switch (jie.dwPOV)
-        {
-            case JOY_POVBACKWARD:
-                if( !JoystickF[C].PovDown)
-                {   JoystickChanged( C, 7); }
-
-                JoystickF[C].PovDown = true;
-                JoystickF[C].PovUp = false;
-                JoystickF[C].PovLeft = false;
-                JoystickF[C].PovRight = false;
-JoystickF[C].PovDnLeft = false;
-JoystickF[C].PovDnRight = false;
-JoystickF[C].PovUpLeft = false;
-JoystickF[C].PovUpRight = false;
-                break;
-case 4500:
-if( !JoystickF[C].PovUpRight)
-                {   JoystickChanged( C, 52); }
-JoystickF[C].PovDown = false;
-                JoystickF[C].PovUp = false;
-                JoystickF[C].PovLeft = false;
-                JoystickF[C].PovRight = false;
-JoystickF[C].PovDnLeft = false;
-JoystickF[C].PovDnRight = false;
-JoystickF[C].PovUpLeft = false;
-JoystickF[C].PovUpRight = true;
-break;
-case 13500:
-if( !JoystickF[C].PovDnRight)
-                {   JoystickChanged( C, 50); }
-JoystickF[C].PovDown = false;
-                JoystickF[C].PovUp = false;
-                JoystickF[C].PovLeft = false;
-                JoystickF[C].PovRight = false;
-JoystickF[C].PovDnLeft = false;
-JoystickF[C].PovDnRight = true;
-JoystickF[C].PovUpLeft = false;
-JoystickF[C].PovUpRight = false;
-break;
-case 22500:
-if( !JoystickF[C].PovDnLeft)
-                {   JoystickChanged( C, 49); }
-JoystickF[C].PovDown = false;
-                JoystickF[C].PovUp = false;
-                JoystickF[C].PovLeft = false;
-                JoystickF[C].PovRight = false;
-JoystickF[C].PovDnLeft = true;
-JoystickF[C].PovDnRight = false;
-JoystickF[C].PovUpLeft = false;
-JoystickF[C].PovUpRight = false;
-break;
-case 31500:
-if( !JoystickF[C].PovUpLeft)
-                {   JoystickChanged( C, 51); }
-JoystickF[C].PovDown = false;
-                JoystickF[C].PovUp = false;
-                JoystickF[C].PovLeft = false;
-                JoystickF[C].PovRight = false;
-JoystickF[C].PovDnLeft = false;
-JoystickF[C].PovDnRight = false;
-JoystickF[C].PovUpLeft = true;
-JoystickF[C].PovUpRight = false;
-break;
-
-            case JOY_POVFORWARD:
-                if( !JoystickF[C].PovUp)
-                {   JoystickChanged( C, 6); }
-
-                JoystickF[C].PovDown = false;
-                JoystickF[C].PovUp = true;
-                JoystickF[C].PovLeft = false;
-                JoystickF[C].PovRight = false;
-JoystickF[C].PovDnLeft = false;
-JoystickF[C].PovDnRight = false;
-JoystickF[C].PovUpLeft = false;
-JoystickF[C].PovUpRight = false;
-                break;
-
-            case JOY_POVLEFT:
-                if( !JoystickF[C].PovLeft)
-                {   JoystickChanged( C, 4); }
-
-JoystickF[C].PovDown = false;
-JoystickF[C].PovUp = false;
-JoystickF[C].PovLeft = true;
-JoystickF[C].PovRight = false;
-JoystickF[C].PovDnLeft = false;
-JoystickF[C].PovDnRight = false;
-JoystickF[C].PovUpLeft = false;
-JoystickF[C].PovUpRight = false;
-                break;
-
-            case JOY_POVRIGHT:
-                if( !JoystickF[C].PovRight)
-                {   JoystickChanged( C, 5); }
-
-JoystickF[C].PovDown = false;
-JoystickF[C].PovUp = false;
-JoystickF[C].PovLeft = false;
-JoystickF[C].PovRight = true;
-JoystickF[C].PovDnLeft = false;
-JoystickF[C].PovDnRight = false;
-JoystickF[C].PovUpLeft = false;
-JoystickF[C].PovUpRight = false;
-                break;
-
-            default:
-                JoystickF[C].PovDown = false;
-                JoystickF[C].PovUp = false;
-                JoystickF[C].PovLeft = false;
-                JoystickF[C].PovRight = false;
-JoystickF[C].PovDnLeft = false;
-JoystickF[C].PovDnRight = false;
-JoystickF[C].PovUpLeft = false;
-JoystickF[C].PovUpRight = false;
-                break;
-        }
-
-        for( short B = 0; B != 32; B ++, jie.dwButtons >>= 1)
-        if( (jie.dwButtons&1))
+if (!Joystick[C].Attached)
+continue;
+
+if (Joystick[C].Left && !JoystickF[C].Left) JoystickChanged(C, 0);
+if (Joystick[C].Right && !JoystickF[C].Right) JoystickChanged(C, 1);
+if (Joystick[C].Up && !JoystickF[C].Up) JoystickChanged(C, 2);
+if (Joystick[C].Down && !JoystickF[C].Down) JoystickChanged(C, 3);
+
+JoystickF[C].Left = Joystick[C].Left;
+JoystickF[C].Right = Joystick[C].Right;
+JoystickF[C].Up = Joystick[C].Up;
+JoystickF[C].Down = Joystick[C].Down;
+
+if (Joystick[C].PovLeft && !JoystickF[C].PovLeft) JoystickChanged(C, 4);
+if (Joystick[C].PovRight && !JoystickF[C].PovRight) JoystickChanged(C, 5);
+if (Joystick[C].PovUp && !JoystickF[C].PovUp) JoystickChanged(C, 6);
+if (Joystick[C].PovDown && !JoystickF[C].PovDown) JoystickChanged(C, 7);
+if (Joystick[C].PovDnLeft && !JoystickF[C].PovDnLeft) JoystickChanged(C, 49);
+if (Joystick[C].PovDnRight && !JoystickF[C].PovDnRight) JoystickChanged(C, 50);
+if (Joystick[C].PovUpLeft && !JoystickF[C].PovUpLeft) JoystickChanged(C, 51);
+if (Joystick[C].PovUpRight && !JoystickF[C].PovUpRight) JoystickChanged(C, 52);
+
+JoystickF[C].PovLeft = Joystick[C].PovLeft;
+JoystickF[C].PovRight = Joystick[C].PovRight;
+JoystickF[C].PovUp = Joystick[C].PovUp;
+JoystickF[C].PovDown = Joystick[C].PovDown;
+JoystickF[C].PovDnLeft = Joystick[C].PovDnLeft;
+JoystickF[C].PovDnRight = Joystick[C].PovDnRight;
+JoystickF[C].PovUpLeft = Joystick[C].PovUpLeft;
+JoystickF[C].PovUpRight = Joystick[C].PovUpRight;
+
+if (Joystick[C].ZUp && !JoystickF[C].ZUp) JoystickChanged(C, 41);
+if (Joystick[C].ZDown && !JoystickF[C].ZDown) JoystickChanged(C, 42);
+if (Joystick[C].RUp && !JoystickF[C].RUp) JoystickChanged(C, 43);
+if (Joystick[C].RDown && !JoystickF[C].RDown) JoystickChanged(C, 44);
+if (Joystick[C].UUp && !JoystickF[C].UUp) JoystickChanged(C, 45);
+if (Joystick[C].UDown && !JoystickF[C].UDown) JoystickChanged(C, 46);
+if (Joystick[C].VUp && !JoystickF[C].VUp) JoystickChanged(C, 47);
+if (Joystick[C].VDown && !JoystickF[C].VDown) JoystickChanged(C, 48);
+
+JoystickF[C].ZUp = Joystick[C].ZUp;
+JoystickF[C].ZDown = Joystick[C].ZDown;
+JoystickF[C].RUp = Joystick[C].RUp;
+JoystickF[C].RDown = Joystick[C].RDown;
+JoystickF[C].UUp = Joystick[C].UUp;
+JoystickF[C].UDown = Joystick[C].UDown;
+JoystickF[C].VUp = Joystick[C].VUp;
+JoystickF[C].VDown = Joystick[C].VDown;
+
+        for( short B = 0; B != 32; B ++)
+        if( Joystick[C].Button[B] )
         {
             if( !JoystickF[C].Button[B])
             {
